src/box/iproto.cc

```diff
@@ -2176,29 +2158,138 @@ iproto_session_push(struct session *session, struct port *port)
+	iproto_thread->push_route[0] =
+		{ iproto_process_push, &iproto_thread->tx_pipe };
+	iproto_thread->push_route[1] = { tx_end_push, NULL };
+	/* IPROTO_OK */
+	iproto_thread->dml_route[0] = NULL;
+	/* IPROTO_SELECT */
+	iproto_thread->dml_route[1] = iproto_thread->select_route;
+	/* IPROTO_INSERT */
+	iproto_thread->dml_route[2] = iproto_thread->process1_route;
+	/* IPROTO_REPLACE */
+	iproto_thread->dml_route[3] = iproto_thread->process1_route;
+	/* IPROTO_UPDATE */
+	iproto_thread->dml_route[4] = iproto_thread->process1_route;
+	/* IPROTO_DELETE */
+	iproto_thread->dml_route[5] = iproto_thread->process1_route;
+	 /* IPROTO_CALL_16 */
+	iproto_thread->dml_route[6] =  iproto_thread->call_route;
+	/* IPROTO_AUTH */
+	iproto_thread->dml_route[7] = iproto_thread->misc_route;
+	/* IPROTO_EVAL */
+	iproto_thread->dml_route[8] = iproto_thread->call_route;
+	/* IPROTO_UPSERT */
+	iproto_thread->dml_route[9] = iproto_thread->process1_route;
+	/* IPROTO_CALL */
+	iproto_thread->dml_route[10] = iproto_thread->call_route;
+	/* IPROTO_EXECUTE */
+	iproto_thread->dml_route[11] = iproto_thread->sql_route;
+	/* IPROTO_NOP */
+	iproto_thread->dml_route[12] = NULL;
+	/* IPROTO_PREPARE */
+	iproto_thread->dml_route[13] = iproto_thread->sql_route;
+	iproto_thread->connect_route[0] =
+		{ tx_process_connect, &iproto_thread->net_pipe };
+	iproto_thread->connect_route[1] = { net_send_greeting, NULL };
+
+	slab_cache_create(&iproto_thread->net_slabc, &runtime);
+	/* Init statistics counter */
+	iproto_thread->rmean = rmean_new(rmean_net_strings, IPROTO_LAST);
+
+	if (iproto_thread->rmean == NULL) {
+		slab_cache_destroy(&iproto_thread->net_slabc);
+		diag_set(OutOfMemory, sizeof(struct rmean),
+			  "rmean_new", "struct rmean");
+		return -1;
+	}
+
+	return 0;
+};
+
 /** Initialize the iproto subsystem and start network io thread */
 void
-iproto_init(void)
-{
-	slab_cache_create(&net_slabc, &runtime);
+iproto_init(unsigned int threads_count)
+{
+	bool fail = false;
+	unsigned int i;
+	iproto_threads_count = 0;
+
+	iproto_threads = (struct iproto_thread *)
+		calloc(threads_count, sizeof(struct iproto_thread));
+	if (iproto_threads == NULL) {
+		tnt_raise(OutOfMemory, threads_count *
+			  sizeof(struct iproto_thread), "calloc",
+			  "struct iproto_thread");
+	}
 
-	/* Init statistics counter */
-	rmean_net = rmean_new(rmean_net_strings, IPROTO_LAST);
-	if (rmean_net == NULL) {
-		slab_cache_destroy(&net_slabc);
-		tnt_raise(OutOfMemory, sizeof(struct rmean),
-			  "rmean", "struct rmean");
+	for (i = 0; i < threads_count; i++, iproto_threads_count++) {
+		struct iproto_thread *iproto_thread = &iproto_threads[i];
+		iproto_thread->number = i;
+		if (iproto_thread_init_routes(iproto_thread) != 0) {
+			fail = true;
+			break;
+		}
+		if (cord_costart(&iproto_thread->net_cord, "iproto",
+				 net_cord_f, iproto_thread)) {
+			rmean_delete(iproto_thread->rmean);
+			slab_cache_destroy(&iproto_thread->net_slabc);
+			fail = true;
```

Comment:
Consider simple refactoring: instead of fail bool, as a rule we use fail label + goto:
```
diff --git a/src/box/iproto.cc b/src/box/iproto.cc
index 236b70cf4..78854ae56 100644
--- a/src/box/iproto.cc
+++ b/src/box/iproto.cc
@@ -2249,6 +2249,11 @@ iproto_init(unsigned int threads_count)
        bool fail = false;
        unsigned int i;
        iproto_threads_count = 0;
+       struct session_vtab iproto_session_vtab = {
+               /* .push = */ iproto_session_push,
+               /* .fd = */ iproto_session_fd,
+               /* .sync = */ iproto_session_sync,
+       };
 
        iproto_threads = (struct iproto_thread *)
                calloc(threads_count, sizeof(struct iproto_thread));
@@ -2261,16 +2266,13 @@ iproto_init(unsigned int threads_count)
        for (i = 0; i < threads_count; i++, iproto_threads_count++) {
                struct iproto_thread *iproto_thread = &iproto_threads[i];
                iproto_thread->number = i;
-               if (iproto_thread_init_routes(iproto_thread) != 0) {
-                       fail = true;
-                       break;
-               }
+               if (iproto_thread_init_routes(iproto_thread) != 0)
+                       goto fail;
                if (cord_costart(&iproto_thread->net_cord, "iproto",
                                 net_cord_f, iproto_thread)) {
                        rmean_delete(iproto_thread->rmean);
                        slab_cache_destroy(&iproto_thread->net_slabc);
-                       fail = true;
-                       break;
+                       goto fail;
                }
                /* Create a pipe to "net" thread. */
                iproto_thread->stopped_connections =
@@ -2285,17 +2287,11 @@ iproto_init(unsigned int threads_count)
                                    iproto_msg_max / 2);
        }
 
-       if (fail) {
-               iproto_free();
-               diag_raise();
-       }
-
-       struct session_vtab iproto_session_vtab = {
-               /* .push = */ iproto_session_push,
-               /* .fd = */ iproto_session_fd,
-               /* .sync = */ iproto_session_sync,
-       };
        session_vtab_registry[SESSION_TYPE_BINARY] = iproto_session_vtab;
+       return;
+fail:
+       iproto_free();
+       diag_raise();
 }

```

Reply:
Yes fixed, you code is better